src/libre/ast_analysis.c

```diff
@@ -552,6 +552,46 @@ set_flags_subtree(struct ast_expr *n, enum ast_flags flags)
 	}
 }
 
+static int
+can_consume_single_newline(struct ast_expr *n)
+{
+	if (!can_consume_input(n)) { return 0; }
+
+	if (n->flags & AST_FLAG_MATCHES_1NEWLINE) { return 1; }
+
+	switch (n->type) {
+	case AST_EXPR_LITERAL:
+		return n->u.literal.c == '\n';
+
+	case AST_EXPR_CODEPOINT:
+		return n->u.codepoint.u == (uint32_t)'\n';
+
+	case AST_EXPR_RANGE:
+		if ((n->u.range.from.type == AST_ENDPOINT_LITERAL) &&
+		    (n->u.range.to.type == AST_ENDPOINT_LITERAL)) {
+			return n->u.range.from.u.literal.c <= '\n'
+			    && n->u.range.to.u.literal.c >= '\n';
+		} else if ((n->u.range.from.type == AST_ENDPOINT_CODEPOINT) &&
+		    (n->u.range.to.type == AST_ENDPOINT_CODEPOINT)) {
+			    return n->u.range.from.u.codepoint.u <= '\n'
+				&& n->u.range.to.u.codepoint.u >= '\n';
+		} else if (n->u.range.from.type == AST_ENDPOINT_NAMED) {
+			/* TODO: unreachable? */
```

Comment:
hm. i'm not sure i understand why this would be unreachable. what happens for a class like `:space:`?

Reply:
`./re -rpcre -p '^abc[[:space:]]de$'` still comes in as `AST_ENDPOINT_LITERAL`, if I put a trace here I get:
```
RANGE TYPES [0 - 0]
-- from 0x09, to 0x0d
```

and likewise ```
$ ./re -rpcre -p '^abc[\d]de$'
RANGE TYPES [0 - 0]
-- from 0x30, to 0x39
```
looking at the parser, it should be coming in as a `NAMED_CLASS`, but something may be rewriting it along the way?

I couldn't figure out a way to reach that branch. I think at one point I locally added an assert there, so if the fuzzer found a way it'd blow up and let me know.

Reply:
Hm. If it is unreachable, let's keep an assert there. I _think_ it would be reached by _ranges_ which end in a class:
```
; ./build/bin/re -r native -pl ast  '^abc[x-[:space:]]de$'
/^abc[x-[:space:]]de$/:8: Unsupported operator
```
but apparently we don't support that.

Reply:
Added an assert in f5b13006.

Reply:
Thanks!